Add four Algo TCA models

vi-to opened the pull request:
Models included in the present PR are Benchmark Price Performance, VWAP benchmarks, PWP benchmark and RPM, described in #54. 

* Multiple benchmarks on a `Symbol` are supported.
* Differently from what is currently done elsewhere in the package I have introduced the `side` param as a number (either 1 for "Buy" or -1 for "Sell"), I believe this is useful to avoid redundant conditional statements.
* When `benchmark = "MktBench"`, a scalar price should be provided via `MktData`. This is enforced by `as.numeric(MktData[1, "MktPrice"])`, to keep the flexibility of `xts` objects inputs on which all the other benchmarks rely and avoid to introduce a new param. However, this is arbitrary and if retained should be stressed.
* At the moment the function does not contain any arrival time param that should go with the `PWP.Shares`, as I am figuring how this relates with the full timespan over which other metrics are computed.
* Also, I have tried to standardize the `MktData` param and several error conditions are introduced in order to help users input data accordingly. This happens in the same way it does for other functions of the package. However, I was not able to throw these errors as eventually they pass, but then `xts` throws its owns (solved with e7d1d58). 

Below are various validation examples, with the newly added data `ABC.Rda` that can be found in `data/`. Trades are sampled from actual market data, as suggested by @jaymon0703.

```
set.seed(333)
.blotter <- new.env()

# sampled trades
data(ABC)
inds <- sample(nrow(ABC), 50)
abc.trades <- ABC[inds]
colnames(abc.trades) <- c('TxnPrice', 'TxnQty')
currency('USD')
stock('ABC', currency = 'USD', multiplier = 1, tick_size = 0.01)
initPortf('abcport', symbols = 'ABC')
addTxns('abcport', 'ABC', TxnData = abc.trades)
updatePortf('abcport', 'ABC')

# market data
startTrades <- first(index(abc.trades))
endTrades <- last(index(abc.trades))
ABC <- subset(ABC, index(ABC) >= startTrades & index(ABC) <= endTrades)
colnames(ABC) <- c("MktPrice", "MktVolmn") # required names
```

Usage of the added function (in progress, output displayed here is outdated).

```
benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'VWAP', type = 'Mkt', MktData = ABC)
benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'VWAP', type = 'Txns', MktData = ABC)
# col header for Performance needs to be looked at
# also, i think we should be more descriptive with the type param. It can be interval VWAP,
# full day VWAP, or VWAP to end of day. The interval equates to type=Txns, and full day equates
# to Mkt...but Txns and Mkt is not very descriptive on its own.

benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'MktBench', type = 'Open', MktData = ABC)
benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'MktBench', type = 'Close', MktData = ABC)
# Performance for both variations is identical. Looking at the code it appears we do not derive
# an open and close price from the data...just an open price being the first price in the series
# We should also think about how we let users specify their own price as numeric.

benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'PWP', POV = 0.25, MktData = ABC)
# PWP might need some more work. The PWP Price is the VWAP for the duration it would have taken
# PWP Shares to trade in the market, starting with the arrival time of the algo order.

benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'RPM', MktData = ABC, verbose = TRUE)
benchTradePerf('abcport', 'ABC', side = -1, benchmark = 'RPM', MktData = ABC, verbose = TRUE)
# May be worth mentioning in the docs the qualitative score mappings table?

# output for above tests:
> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'VWAP', type = 'Mkt', MktData = ABC)
$Trades.VWAP.Perf
  Symbol Side Avg.Exec.Price VWAP.Mkt MktVolmn
1    ABC  Buy       4578.195 4630.939 113.8966

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'VWAP', type = 'Txns', MktData = ABC)
$Trades.VWAP.Perf
  Symbol Side Avg.Exec.Price VWAP.Txns               Txn.Qty
1    ABC  Buy       4578.195  4578.195 -0.000000000001986579

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'MktBench', type = 'Open', MktData = ABC)
$Trades.MktBench.Perf
  Symbol Side Avg.Exec.Price MktBench.Open Performance
1    ABC  Buy       4578.195          5018    876.4556

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'MktBench', type = 'Close', MktData = ABC)
$Trades.MktBench.Perf
  Symbol Side Avg.Exec.Price MktBench.Close Performance
1    ABC  Buy       4578.195           5018    876.4556

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'PWP', POV = 0.25, MktData = ABC)
$Trades.PWP.Perf
  Symbol Side Avg.Exec.Price  POV PWP.Shares PWP.Price Performance
1    ABC  Buy       4578.195 0.25    2556120   4549.02   -64.13462

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'RPM', MktData = ABC, verbose = TRUE)
$Trades.RPM.Perf
  Symbol Side Avg.Exec.Price t.Mkt.Volmn t.Fav.Volmn t.Unfav.Volmn       RPM Quality
1    ABC  Buy       4578.195    83470816    47628292      35842524 0.5705981 Average

> benchTradePerf('abcport', 'ABC', side = -1, benchmark = 'RPM', MktData = ABC, verbose = TRUE)
$Trades.RPM.Perf
  Symbol Side Avg.Exec.Price t.Mkt.Volmn t.Fav.Volmn t.Unfav.Volmn       RPM Quality
1    ABC Sell       4578.195    83470816    35842524      47628292 0.4294019 Average

# multiple benchmark tests
benchTradePerf('abcport', 'ABC', side = 1, benchmark = c('MktBench', 'VWAP', 'PWP', 'RPM'), POV = 0.25, type = 'Mkt', MktData = ABC)

# output
> benchTradePerf('abcport', 'ABC', side = 1, benchmark = c('MktBench', 'VWAP', 'PWP', 'RPM'), POV = 0.25, type = 'Mkt', MktData = ABC)
$Trades.MktBench.Perf
  Symbol Side Avg.Exec.Price MktBench.Mkt Performance
1    ABC  Buy       4578.195         5018    876.4556

$Trades.VWAP.Perf
  Symbol Side Avg.Exec.Price VWAP.Mkt MktVolmn
1    ABC  Buy       4578.195 4630.939 113.8966

$Trades.PWP.Perf
  Symbol Side Avg.Exec.Price  POV PWP.Shares PWP.Price Performance
1    ABC  Buy       4578.195 0.25    2556120   4549.02   -64.13462

$Trades.RPM.Perf
  Symbol Side Avg.Exec.Price t.Mkt.Volmn t.Fav.Volmn t.Unfav.Volmn       RPM
1    ABC  Buy       4578.195    83470816    47628292      35842524 0.5705981

```

Double check of the metrics, especially of the PWP benchmark, is sought after. Often there is some little ambiguity on them, also when not accounting for data vendors specific filtering. 

Documentation is kept at a bare minimum at the moment as the models are extensively discussed in the issue.

vi-to committed:
Merge branch 'master' of git://github.com/braverock/blotter into 54_algotca

vi-to committed:
Update fork with last changes

Merge branch '54_algotca' of git://github.com/braverock/blotter into 54_algotca

# Conflicts:
#	DESCRIPTION

vi-to committed:
Update DESCRIPTION dependencies to match 54_algotca branch

vi-to committed:
Merge branch '54_algotca' of git://github.com/braverock/blotter into 54_algotca

vi-to committed:
Fix typos in impShortfall() function

See #54

vi-to committed:
Merge branch '54_algotca' of git://github.com/braverock/blotter into 54_algotca

vi-to committed:
Add four Algo TCA benchmarking models

Models included are Benchmark Price Performance, VWAP benchmarks, PWP benchmark and RPM

See #54

vi-to committed:
Attempt to fix .Rd error in docs

vi-to committed:
Attempt to remove .Rd error in docs

vi-to committed:
Attempt to solve 'no visible binding for global variable: type.convert'

Apparently there was a warning other than .Rd undefined sequence.
Adding an explicit namespace for 'type.convert' resolves it locally

vi-to committed:
Remove repetition in PnL performance calculation

vi-to committed:
Improve benchTradePerf()

Now all errors are thrown as intended, by defining the variable tMktVlmn in place where needed.

Modifications in benchmark = 'RPM' are:
- enhance efficiency of xts extraction
- seek interpretability gains on both devs and users sides, introducing a clearer distinction between working variables and a consistent output display
- update docs accordingly

vi-to committed:
Correct and simplify tFavVolmn and tUnfavVolmn variables, their typos and their output handling

vi-to committed:
Fix tUnfavVolmn variable typo

vi-to commented:
Relevant output example updated to fc5e6ad.

```
> benchTradePerf('abcport', 'abc', side = 1, benchmark = 'RPM', MktData = ABC, verbose = TRUE)
$Trades.RPM.Perf
  Symbol Side Avg.Exec.Price t.Mkt.Volmn t.Fav.Volmn t.Unfav.Volmn       RPM Quality
1    abc  Buy       4578.195    83470816    47628292      35842524 0.5705981 Average
```
**EDIT**: commit a66a41d simplifies coding and handling rather than "correct" in a strict sense.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' Trade Execution Performance Benchmarks
+#' 
+#' This function gathers different benchmarking methods used to evaluate the average execution price \eqn{\bar{P}} of a given trading strategy.
+#' The \eqn{\bar{P}} is compared against a number of benchmark metrics, in order to assess its performance in terms of profit or loss.
```
Thanks Vito, few comments about things we can clean up before this can be merged.

- Try and keep documentation line length limited to 80 characters - https://google.github.io/styleguide/Rguide.xml#linelength
- Benchmarking execution performance using TCA models is not about profit and loss. PnL is itself the ultimate benchmark, but thats a different objective. With TCA we are looking to minimize costs as opposed to maximize portfolio value.

vi-to replied:
Thank you for your observations. 

Although the documentation was not intended to be definitive, I will happily keep in mind this aspect as part of the style you prefer.

Your second point is unclear to me. Each PnL metric provides a (bps) difference between a given execution strategy and a benchmark, which implies a profit or a loss for the trader. I am just writing the benchmarking. Would you please explain further your observation with respect to this function?

jaymon0703 replied:
I guess i would say "The \eqn{\bar{P}} is compared against a number of benchmark metrics, in order to assess its performance in terms of profit or loss relative to the benchmark."

vi-to replied:
I though this was understood by the "compared against..." and the benchmarking context overall, but I gratefully welcome your suggestions to better document _this_.

jaymon0703 replied:
We should also be explicit in the Decription at least by stating that a negative value for performance indicates underperformance, and a positive value outperformance.

vi-to replied:
Sure, it is already stated.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' 
+#' This function gathers different benchmarking methods used to evaluate the average execution price \eqn{\bar{P}} of a given trading strategy.
+#' The \eqn{\bar{P}} is compared against a number of benchmark metrics, in order to assess its performance in terms of profit or loss.
+#' These methods are not strictly alternative, indeed each of them provides different insights and suffers limitations.
```
```suggestion

```

vi-to replied:
Very often they are compared and suggested as alternatives. Are there motivations to delete this?

jaymon0703 replied:
Sorry for not explaining. I thought the language "suffers limitations" was a bit strong. Using the benchmarks in conjunction is definitely best practice. They are not mutually exclusive, agreed. Perhaps referencing https://arxiv.org/pdf/1904.01566.pdf would be useful here.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' The \eqn{\bar{P}} is compared against a number of benchmark metrics, in order to assess its performance in terms of profit or loss.
+#' These methods are not strictly alternative, indeed each of them provides different insights and suffers limitations.
+#' 
+#' The performance is quantified by means of a \emph{Profit and Loss (PnL) metric}. 
```
This could be confusing without specific reference to the formula in Kissel, since PnL generally refers to marking a stock or portfolio to market, or the difference in value once a position is unwound. I suppose we could be explicit about the PnL being referenced against the benchmark.

vi-to replied:
Similarly to what commented elsewhere, yes it will be referenced. It will be also remarked that the RPM does an exception.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' 
+#' The performance is quantified by means of a \emph{Profit and Loss (PnL) metric}. 
+#' A positive PnL metric indicates that the trading strategy outperformed a chosen benchmark on average,
+#' viceversa negative values register an underperformance.
```
```suggestion
#' vice versa negative values register an underperformance.
```

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' 
+#' 
+#' where \eqn{P_{avg}} is the average execution price and \eqn{P_{B}} is a given benchmark price that can be a single next/present open/close price.
+#' It is worth stressing that they are expressed in basis points (bps) units.
```
Did we say we were going to add another unit, in $?

vi-to replied:
I think the units aspect may be better addressed with a separate convenience function.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' 
+#' Lastly the \emph{Relative Performance Measure}, which differs from the PnL metrics above. It is expressed as:
+#' 
+#' 
```
I am guessing this is blank as it is a TODO?

vi-to replied:
Yes. In reality, as I wrote in the PR opening comment all the documentation other than the code related one is a TODO at the moment. Although it will be documented, there is much book text in the issue and a vignette will be written as well.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' @param Portfolio A portfolio name that points to a portfolio object structured with initPortf()
+#' @param Symbol A string identifying the traded symbol to benchmark
+#' @param side Numeric, either 1 or -1, that indicates the side of the trade. \code{side = 1} means "Buy" and \code{side = -1} is "Sell". Default is \code{side = 1}.
+#' @param benchmark A string or vector of strings providing the benchmark metric(s). One of 'MktBench', 'VWAP', 'PWP', 'RPM'. Default is #TODO.
```
why MktBench? wont Bench do? they all involve some or other market benchmark

vi-to replied:
I agree with you on rename it. However a generic `Bench` would incur the same sort, as you said. `Bprice` or perhaps a more consistent `BP`?

vi-to replied:
Updates on this name?

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' @param benchmark A string or vector of strings providing the benchmark metric(s). One of 'MktBench', 'VWAP', 'PWP', 'RPM'. Default is #TODO.
+#' @param type A string specifying the type of the benchmark used. Relevant only for \code{benchmark = 'MktBench'} and \code{benchmark = 'VWAP'}. 
+#'             In the former case it only changes the output format, while in the latter it also specifies the VWAP to be used. If \code{benchmark = 'VWAP'}, default is \code{type = 'Txns'}
+#' @param MktData An xts object containing a 'MktPrice' and 'MktVolmn' required columns.
```
Lets try keep things camelCase, so mktData or mkt.Data (Google R style guidelines prefer the latter but the former is more pervasive throughout blotter, so i would go with that for consistency). However, there is a `mktdata` object in quantstrat which we do not want to get confused with...assuming they could ever exist in the save environment which they probably could. Naming things is hard, as they say.

Also, why `Volmn`?

We should try and use the colname requirements in addTxns...namely: `TxnQty` and `TxnPrice`. From addTxns documentation - `The TxnData object must have "TxnQty" and "TxnPrice" columns, while the "TxnFees" column is optional`

vi-to replied:
Indeed I think `MktData` is consistent with `TxnData` in `addTxns`, for example. 

I didn't like much `.Volmn` either. If we agree it will be turned in `.Qty` all around.
I propose to keep the `Mkt` distinction to avoid confusion with transaction quantities.

vi-to replied:
So keeping `MktData` param and going ahead with `MktQty` edits?
It overall seems to integrate well with corresponding transactions naming conventions of the package.

jaymon0703 replied:
agree with Qty

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' \describe{
+#'      \item{\code{Symbol}: }{A string identifying the traded symbol to benchmark}
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell".}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
```
camelCase
```suggestion
#'      \item{\code{avg.Exec.Price}: }{Symbol transactions average execution price}
```

vi-to replied:
This is a console output name

vi-to replied:
Most console output columns of the functions in the package have first capital letter and dot naming convention. Therefore resolving.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell".}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{MktBench.*}: }{The benchmark and an arbitrary \code{type} provided as input (e.g. 'Open', 'Close').}
+#'      \item{\code{Performance}: }{The \emph{Benchmark PnL} performance, in bps.}
```
No need for full stops when documenting parameters, unless of course you use 2 sentences
```suggestion
#'      \item{\code{Performance}: }{The \emph{Benchmark PnL} performance, in bps}
```

vi-to replied:
Thank you for this documentation style suggestion

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#'      \item{\code{Symbol}: }{A string identifying the traded symbol to benchmark}
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell".}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{VWAP.*}: }{The benchmark and depending on \code{type} parameter either 'Txns' 'Mkt'.}
```
we use "txns" elsewhere in blotter so lets aim for consistency
```suggestion
#'      \item{\code{VWAP.*}: }{The benchmark and depending on \code{type} parameter either 'txns' or 'mkt'.}
```

vi-to replied:
It's just a character param. I wrote it with a first capital letter only since as you will see is used for console output.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#'      \item{\code{Side}: }{The \code{side} of the trades, as "Buy" or "Sell".}
+#'      \item{\code{Avg.Exec.Price}: }{Symbol transactions average execution price}
+#'      \item{\code{t.Mkt.Volmn}: }{Total market volume over the order timespan.}
+#'      \item{\code{t.Fav.Volmn}: }{Total market volume over the order timespan for which the average execution price of a 'Buy' ('Sell') order was lower (greater) then market prices.}
```
```suggestion
#'      \item{\code{t.Fav.Volmn}: }{Total market volume over the order timespan for which the average execution price of a 'Buy' ('Sell') order was lower (greater) than market prices.}
```

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#' @param side Numeric, either 1 or -1, that indicates the side of the trade. \code{side = 1} means "Buy" and \code{side = -1} is "Sell". Default is \code{side = 1}.
+#' @param benchmark A string or vector of strings providing the benchmark metric(s). One of 'MktBench', 'VWAP', 'PWP', 'RPM'. Default is #TODO.
+#' @param type A string specifying the type of the benchmark used. Relevant only for \code{benchmark = 'MktBench'} and \code{benchmark = 'VWAP'}. 
+#'             In the former case it only changes the output format, while in the latter it also specifies the VWAP to be used. If \code{benchmark = 'VWAP'}, default is \code{type = 'Txns'}
```
This probably needs more explanation, which can be covered in a `Details` section. You can specify here to `see Details`

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,200 @@
+#'             In the former case it only changes the output format, while in the latter it also specifies the VWAP to be used. If \code{benchmark = 'VWAP'}, default is \code{type = 'Txns'}
+#' @param MktData An xts object containing a 'MktPrice' and 'MktVolmn' required columns.
+#' @param POV A number between 0 and 1, specifying the POV rate.
+#' @param verbose A logical. It allows a RPM qualitative score appended. Default is \code{FALSE}.
```
```suggestion
#' @param verbose A logical. It allows a RPM qualitative score to be appended. Default is \code{FALSE}.
```

jaymon0703 commented:
Thanks for the example...we should probably make sure our sampled transactions are in order before updating the portfolio just so it resembles reality.
![image](https://user-images.githubusercontent.com/4340156/58907676-09315d00-870f-11e9-8664-003e6c21e594.png)

vi-to commented:
Hi @jaymon0703, what is that `txns` object? The `abc.trades` in the example are ordered.

jaymon0703 reviewed (commented):
Great first PR to kick start this GSoC 2019 project. Thanks @vi-to. Please see my comments and suggestions. Its getting late and i have had a long day, so i will review the test results and code structure specifically tomorrow.

jaymon0703 commented:
> Hi @jaymon0703, what is that `txns` object? The `abc.trades` in the example are ordered.

Weird, its not ordered for me, and i dont see how it possibly would be ordered since inds will be randomly sampled indexes. The `txns` object was inside the function.

vi-to commented:
A simple answer is that we are subsetting `xts` objects. They are time-based objs and their subsetting mechanism is a time-based queries one, which ultimately should refer to `.parseISO8601` and the standard behind. 

The `txns` object in the function is produced by other blotter functions in the `@seealso`. I did not experience any particular problem. From your screenshot it seems dates are increasingly ordered, do you think there is a problem with them?

jaymon0703 commented:
i will note to have another look during testing today

vi-to committed:
Update benchTradePerf() docs

- fix some typos
- remove full stops where requested
- refine some points
- add some TODOs where formulas are needed

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,214 @@
+    
+    # PnL performance for 'MktBench', 'VWAP' (all types) and 'PWP' 
+    if (benchmark[i] != 'RPM') {
+      out[, 'Performance'] <- (-1) * side * (p_avg - benchPrice)/benchPrice * 10000
```
For some reason, with benchTradePerf(VWAP) the Performance column in the output has the name MktVolmn

vi-to replied:
This really should not happen. Similarly, in the editing you have done on my original PR comment I have noticed that a new example with `type = 'Txns'` was introduced and that it shows a `Txn.Qty` column instead of `Performance`. But if you check the object, the name is correctly 'Performance' in both cases.

jaymon0703 replied:
i think i see the problem...benchPrice is a matrix result from crossprod()...if you use benchPrice[1,1] it works. Alternatively as opposed to using crossprod() just use cumsum*cumsum?

jaymon0703 replied:
Yes, the new example test with type='Txns' was to cover the other scenario

vi-to replied:
Great solution. However, it is a curious error, `pwpPrice` goes the same way but it does not fail in giving the correct 'Performance' column. I found a different solution which produces correct results and it is perhaps easier to maintain.

vi-to replied:
Solved with f7eea99

jaymon0703 replied:
happy

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,214 @@
+                           Symbol,
+                           side = 1,
+                           benchmark = c("MktBench", "VWAP", "PWP", "RPM"),
+                           type = c("Txns", "Mkt"),
```
I think we should be more descriptive with type, interval seems to be Txns, full_day seems to be Mkt...on their own Txns and Mkt are not very self explanatory. Interval VWAP and Full Day VWAP are fairly commonly used terms.

vi-to replied:
I agree with you, there is a bit of ambiguity around 'interval' and 'full' periods. Since this also points to the PWP arrival time, I suggest to gather these discussions. A couple of new params for dates may probably follow.

jaymon0703 replied:
Unfortunately i cannot do multi-line suggestions in reviewing a PR, so will post the full function code here. You can do a diff using BeyondCompare or some equivalent to see what is different, but i have simply added subsetting to interval.VWAP, besides replacing type=("Txns","Mkt") with type=("interval.VWAP","full.day.VWAP") in the function definition

PWP will require similar subsetting...but i will only be able to provide a suggestion this evening

```
benchTradePerf <- function(Portfolio,
                           Symbol,
                           side = 1,
                           benchmark = c("MktBench", "VWAP", "PWP", "RPM"),
                           type = c("interval.VWAP", "full.day.VWAP"),
                           MktData,
                           POV = NULL,
                           verbose = FALSE
                           )
{ 
  pname <- Portfolio
  Portfolio <- .getPortfolio(pname)
  txns <- Portfolio[["symbols"]][[Symbol]][["txn"]]
  p_avg <- as.numeric(last(txns$Pos.Avg.Cost))
  tTxnQty <- sum(txns$Txn.Qty)
  
  # Benchmark metrics
  benchmark <- match.arg(benchmark, c("MktBench", "VWAP", "PWP", "RPM"), several.ok = TRUE)
  
  tradesPerf <- list()
  
  for (i in 1:length(benchmark)) {
    switch (benchmark[i],
            MktBench = {
              if (!("MktPrice" %in% colnames(MktData))) stop("No MktPrice column found, what did you call it?")
              if (missing(type)) type <- NULL
              
              benchPrice <- as.numeric(MktData[1, "MktPrice"])
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
            },
            VWAP = {# full VWAP
              if (missing(type)) type <- 'interval.VWAP'
              
              if (type == 'interval.VWAP') { # interval VWAP
                inds <- which(strftime(index(MktData), format = "%H:%M:%S", tz = "UTC") == strftime(index(txns), format = "%H:%M:%S", tz = "UTC"))
                inds <- sort(inds)
                start.idx <- first(inds)
                end.idx <- last(inds)
                benchPrice <- crossprod(MktData$MktPrice[start.idx:end.idx], MktData$MktVolmn[start.idx:end.idx])/sum(MktData$MktVolmn[start.idx:end.idx])
                } else if (type == 'full.day.VWAP') { # full.day.VWAP
                  if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
                  benchPrice <- crossprod(MktData[, "MktPrice"], MktData[, "MktVolmn"])/sum(MktData[, "MktVolmn"])
                }
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
            },
            PWP = {
              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
              # if (missing(arrTime)) arrTime <- first(index(txns))
              
              pwpShares <- rep(tTxnQty/POV, length(txns$Txn.Price))
              benchPrice  <- crossprod(txns$Txn.Price, pwpShares)/sum(pwpShares) # PWP price
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
            },
            RPM = {
              if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
              
              tMktVolmn <- sum(MktData[, "MktVolmn"])
              
              if (side == 1) {
                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
              } else {
                tFavVolmn   <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] < p_avg])
                tUnfavVolmn <- sum(MktData[, "MktVolmn"][MktData[, "MktPrice"] > p_avg])
              }
              
              rpm <- 0.5*(tMktVolmn + tFavVolmn - tUnfavVolmn)/tMktVolmn
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, tMktVolmn, tFavVolmn, tUnfavVolmn, rpm), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 't.Mkt.Volmn', 't.Fav.Volmn', 't.Unfav.Volmn', benchmark[i])
              
              # Append RPM qualitative score
              if (verbose) { 
                if (rpm >= 0.4 & rpm <= 0.6) {
                  quality <- "Average"
                } else if (rpm > 0.6) {
                  quality <- "Good"
                  if (rpm >= 0.8) {
                    quality <- "Excellent" 
                  }
                } else {
                  quality <- "Fair"
                  if (rpm < 0.2) {
                    quality <- "Poor"
                  }
                }
                out[, 'Quality'] <- quality 
              }
            }
    )
    
    # PnL performance for 'MktBench', 'VWAP' (all types) and 'PWP' 
    if (benchmark[i] != 'RPM') {
      benchPrice <- as.numeric(benchPrice)
      out[, 'Performance'] <- (-1) * side * (p_avg - benchPrice)/benchPrice * 10000
    }
    
    # Store and preserve data types
    tradesPerf[[i]] <- utils::type.convert(out, as.is = TRUE)
    names(tradesPerf)[i] <- paste('Trades', benchmark[i], 'Perf', sep = '.')
    row.names(tradesPerf[[i]]) <- NULL
  }
  return(tradesPerf)
}

```

vi-to replied:
Thank you @jaymon0703. Started working on VWAP modifications.

* With your 'interval.VWAP' you introduced a different way to calculate the VWAP, which from the code is referred to market quantities as for your 'full.day.VWAP'. I found that VWAP is sometimes calculated with respect to transactional data as opposed to market one and `type` was used accordingly to account for this distinction and not for the timespan. If this is a correction, I would be thankful if you could please confirm and elaborate it a little more, as this is precisely the ambiguity I have been asking for clarification. Perhaps also @braverock may want to add some remarks here.
* The coding style of variables you introduced is different from the camelCase requested. Assuming I should keep the latter.
* Why 'full.day'? There could be a very different timing, you did not change anything else in the `benchPrice`. May 'full' be more appropriate?

Waiting for PWP details, so I can commit these time-based modifications.

jaymon0703 replied:
- The volume weighted average price of all trades on the market *during* the time that your algo (or manual) order is trading is considered a fair price and is the general benchmark used when benchmarking execution performance with a VWAP benchmark. VWAP benchmarking means benchmarking your average execution price to interval VWAP (ie. the VWAP between the start and stop time of your order). If you compare your average execution to full day VWAP or VWAP to End of Day, you include an opportunity cost component of not trading.  
- intervalVWAP is fine but full.dayVWAP feels wrong. Google's styling code prefers camelCase with periods, but naming things is hard so whatever you feel comfortable with taking all this into account...as long as the function works i dont think it matters too much.
- the assumption is user has added full day's trades and Kissell refers to full day, which makes sense to compare the fair price for the entire day. If user does not have full day's trades, then full day VWAP will not be appropriate...and user will not use type="full day". This is why documentation will be impotant.

vi-to replied:
* Time dimension acquired, the remaining question is: Is the VWAP always computed from market data, i.e. is it always the weighted average of market prices on market volume? 
* Pointed to `start.idx` and `end.idx`, but got your point... I will try to keep a consistent style if possible.
* Agree, but perhaps there is no need to include '.day'. If you input daily data you know that you are having a daily calc and the same goes for other timing. This usage of the function can be vignette examples?

jaymon0703 replied:
Regarding your question: "Time dimension acquired, the remaining question is: Is the VWAP always computed from market data, i.e. is it always the weighted average of market prices on market volume?"

The answer is yes.

I will resolve this conversation if you agree.

vi-to replied:
Thank you @jaymon0703. Once the last PWP details on arrival time are discussed I can figure out if introducing a new time parameter of the function can suitably fit all of the above needs or if more of them are appropriate. Commenting about it in the review comment you opened.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,214 @@
+    switch (benchmark[i],
+            MktBench = {
+              if (!("MktPrice" %in% colnames(MktData))) stop("No MktPrice column found, what did you call it?")
+              mktBenchPrice <- as.numeric(MktData[1, "MktPrice"])
```
does this mean mktBenchPrice will always be the first price in the series? should we account for last price in the series too? should we give user option for inputting their own price?

vi-to replied:
No, it shouldn't have to in general. Yes, it should be really a price provided by the user.

vi-to replied:
This aspect is being extensively discussed in the main PR conversation.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,214 @@
+              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
+              # if (missing(arrTime)) arrTime <- first(index(txns))
+              
+              pwpShares <- rep(tTxnQty/POV, length(txns$Txn.Price))
```
PWP might need some more work. The PWP Price is the VWAP for the duration it would have taken
PWP Shares to trade in the market, starting with the arrival time of the algo order.

vi-to replied:
See https://github.com/braverock/blotter/pull/93#discussion_r290834253

vi-to replied:
So here as well PWP price needs to be based on market prices. 
Also with respect to the arrival time, as we are assuming all around transactions added to the symbol are the order, thus arrival time would simply be the first transaction time. Hence no need for an explicit new time param here, right?

jaymon0703 replied:
Assuming Arrival Time is equivalent to when the first trade occurred is sufficient for now. As discussed, Arrival Time is when your order hits the order book, which may be minutes or longer before the first trade time, and would impact the market so is a more accurate starting timestamp to begin benchmarking.

I will make code suggestion for PWP shortly. Since multi-line suggestions are not possible in GitHub i will likely just post snippet of PWP code here.

jaymon0703 replied:
Below is my PWP code suggestion. This needs to be tested, but is identical to my suggestion for VWAP except that end.idx is not the index of the final txn used for calculating our VWAP, but the index of the final market txn where cumsum(txns) equates to PWP Shares.

```
            PWP = {
              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
              # if (missing(arrTime)) arrTime <- first(index(txns))
              
              pwpShares <- tTxnQty/POV
              inds <- which(strftime(index(MktData), format = "%H:%M:%S", tz = "UTC") == strftime(index(txns), format = "%H:%M:%S", tz = "UTC"))
              inds <- sort(inds)
              start.idx <- first(inds)
              # now we need the final timestamp at which cumsum volume equates to PWP Shares...this wont be exact, but we will get close
              # we will use findInterval() for this...i love findInterval() 
              end.idx <- findInterval(pwpShares, cumsum(MktData$MktVolmn[start.idx:nrow(MktData)])) # end.idx is the final timestamp at which cumsum volume equates to PWP Shares
              benchPrice <- crossprod(MktData$MktPrice[start.idx:end.idx], MktData$MktVolmn[start.idx:end.idx])/sum(MktData$MktVolmn[start.idx:end.idx]) # PWP price
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
            },
            RPM = {
```

vi-to replied:
Thank you @jaymon0703, your clarifications really helped me demystify some tricky definitions.
There is however something that may be perhaps an unintended behaviour of `findInterval()` in our context. To give an explanatory example, say our `pwpShares` are 14500. Using our testing data and all the rest being the same:

```
> ind <– findInterval(14500, cumsum(ABC[, "MktVolmn"]))
> ind
 [1] 7
```

but at this time

```
> cumsum(ABC[, "MktVolmn"])[ind]
                    MktVolmn
2019-02-01 09:04:46     4104
```
which is considerably lower than our 14500 shares. In fact, 14500 is much closer to the next observed cumulative market volume, 14882. Results are the same until this last value is reached.

I have tested that function with its different options, but the result was the same. Perhaps I left out one of its aspects and we can indeed use it, please point them out. 

In the meanwhile, I am committing an alternative solution for you to consider. It is still a rough approximation but more often we will get closer.

jaymon0703 replied:
The typical order duration someone would want to benchmark will be at least 60mins. findInterval() in my suggestion finds the index at which cumsum(MktData) is closest to PWP Shares. Adding the volume at the next index will take us above PWP Shares. What this means is that we never get to PWP Shares exactly when computing PWP benchmark...but we get close...and for a realistic order with duration 60+mins the impact of the next trade on the PWP Price will be negligible.

In summary, i would not worry about not getting to PWP Shares exactly, this is impossible to do...unless we use the portion of the volume at the next index which makes our cumsum(MktData) == PWP Shares. This will be exact and theoretically more correct, but the benefit over my suggestion should be negligible for typical orders which will be benchmarked in production (ie. 60+mins).

vi-to replied:
Agree on the added complexity needed, but the solution I am about to propose is very simple and perhaps it is worth it in this phase.

jaymon0703 commented on R/benchTradePerf.R:
```diff
@@ -0,0 +1,214 @@
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
+            },
+            RPM = {
```
May be worth mentioning in the docs the qualitative score mappings table?

RPM | Quality |
------------ | ------------- |
0-20% | Poor |
20-40% | Fair |
40-60% | Average |
60-80% | Good |
80-100% | Excellent |

vi-to replied:
Why not, although please note that I have decided (balanced) bounds which are not stated in the issue.

jaymon0703 replied:
cool, we cannot have bounds overlapping

vi-to committed:
Fix and clean benchTradePerf()

- fix columns names overflow in console output for benchmark = 'VWAP', caused by 'benchPrice' class
- use 'benchPrice' common variable to avoid repetitive code (it now definitively replaces 'mktBenchPrice', 'vwapPrice' and 'pwpPrice')
- concisely preserve output data types

vi-to committed:
Avoid a missing 'type' to break the function when benchamark = 'MktBench'

vi-to commented:
@jaymon0703 commenting here to reply about some notes you left in https://github.com/braverock/blotter/pull/93#issue-284375836 edits and gather many sparse comments in the reviews.

> Performance for both variations is identical. Looking at the code it appears we do not derive
   an open and close price from the data...just an open price being the first price in the series.

Yes, the performance is and should be identical as long as you are providing the same `MktData[,"MktPrice"]`. An user is free to input prices of his choice, which is his data. The `type` param when `benchmark = 'MktBench'` is simply used to customize the name of the benchmark column in the console output. This was touched on in the documentation, but will be stressed again. 

Also, working on allowing a numeric value. I would like to keep `xts` objects. Because of this I propose to include it as well in our very next discussion on VWAP intervals and PWP arrival time (see https://github.com/braverock/blotter/pull/93#discussion_r290834253).

Apart from documentation, which I am keeping in mind, are there other aspects you think are left behind?

jaymon0703 commented:
I would have expected type='Open' to give a different performance output when compared with type='Close'. It does not appear that in the code we differentiate the output for MktBench based on the type the user specifies. We always set mktBenchPrice = as.numeric(MktData[1, "MktPrice"]) regardless of type. Should we not use `as.numeric(MktData[nrow(MktData), "MktPrice"])` when user specifies `type='Close'`?

vi-to commented:
I doubt there is a way to deduce which kind of prices someone inputs in `MktData[, "MktPrice"]` and this is why that column name is kept generic. It may represent any of the OHLC prices or a vector of prices of your choice for all that matters, say a next day open price. I am the first one not knowing what kind of prices are in the `ABC.Rda` I have used. It is up to the analyst and his research. The `type` does not influence that, as the corresponding PnL metric is computed regardless of it. As said, I felt providing this parameter also when `benchmark='MktBench'` was a suitable way to have the output column name customized with the kind of prices you have chosen (which only you know) and distinguish each other from other kinds you could be including in your analysis (you may be benchmarking against both next day open prices and next day closing prices, providing both of them separately in each call to the function. Then to which `MktBench` are you referring to?).

As long as you have separate columns for open prices and closing prices, I am not sure `as.numeric(MktData[nrow(MktData), "MktPrice"])` on open prices will happen to give you a closing price. 
To have an example, let's say `MktData[, "MktPrice"]` are your closing prices. You want to benchmark your execution against one single value of them?

```
> single.close.price <- ABC[2019, ] # benchmark price of your choice, used 2019 only to give the example

> benchTradePerf('abcport', 'ABC', side = 1, benchmark = 'MktBench', type = 'Close', MktData = single.close.price)
$Trades.MktBench.Perf
  Symbol Side Avg.Exec.Price MktBench.Close Performance
1    abc  Buy       4578.195           4914    683.3647
``` 

This allows to keep `xts` objects consistently all over the function. In fact, similarly to many other judicious package functions, `benchTradePerf()` is coded on the assumption that a user has already built a portfolio and added its transactional data to it through this kind of objects.

jaymon0703 commented:
Well my expectation is for the Benchmark Price Performance metric to be useful you probably want to compare your average execution price to an Open price and a Close price, both of which can be deduced from the first and last observation in your market data object, assuming the market data object is the full day's trades.

vi-to commented:
Yes, it could be. But again, are you benchmarking your daily transactions against the daily closing? Are you benchmarking your order (what we are actually doing with `p_avg`) against the last closing price referred to all of them and possibly spanning on many different trading days (as `ABC.Rda` data is)? 

What I am trying to say is that with that input you can subset the `MktData` input freely and consequently have the benchmark you need.

jaymon0703 commented:
The Open and Close in Kissell refers to the current day Open and Close. I think its feasible to expect a user to specify a MktData object of the full day's trades, from which an Open and Close can be deduced.

vi-to commented:
What I am finding difficult to understand from your comments is why such a restriction should be introduced. Are entire orders with transactions over several days comparable with one single closing price? Which one of the many?

In case you have daily trades only and your `MktData` is over the day:
```
single.close.price <- ABC[nrow(ABC), ] # your request above
benchTradePerf('abcport', 'abc', side = 1, benchmark = 'MktBench', type = 'Close', MktData = single.close.price)
```
Any other else would go similarly.

Is there something I am missing?

jaymon0703 commented:
I guess i am confused by mentioning Open and Close in the docs but not accounting for it in the code.

Orders for benchmarking with TCA most often do not span longer than 1 day. Using the totals day's trades as MktData will be sufficient for any algo order started and finished on the same day, which 99% will be.

vi-to committed:
Update benchTradePerf() docs and fix one RPM quality bound consistently

vi-to committed:
Add more docs to benchTradePerf()

Documentation added is on benchamarks considerations.
Also trying to adhere to the 80 characters lines convention

vi-to committed:
Remove a duplicated RPM concept in docs

The concept removed was better explained with documentation addition in b12030

vi-to committed:
Improve benchTradePerf() behavior for benchmark='MktBench'

- add parameter 'priceToBench' to select a price in 'MktData'
- add possibility to input a single numeric value through 'MktData'

In both cases the resulting price is the benchmark price of choice

vi-to committed:
Change 'type' related benchTradePerf() structure

- 'type' parameter is now a list, this helps giving structure to the function and separate its different usages throughout
- correct and adapt benchmark='VWAP'

vi-to committed:
Fix benchTradePerf() when benchmark='PWP'

Correct PWP calcs. Note that when PWP shares are reached in terms of cumulative market volume is still an approximation

vi-to committed:
Complete commit 5205d2b988def8a30f58f50bb11eddb70db5a429

They were intended to be one and the same. Namely, to introduce a different control flow through the 'type' parameter, fix and conform benchmark='VWAP' accordingly.
Minor docs modifications followed

vi-to committed:
Enrich benchTradePerf() docs

Added VWAP types coded in the function and mainly discussed most function usages details

vi-to committed:
Fix missing 'type' fallback for benchmark='MktBench'

vi-to committed:
Apply naming convention, 'MktVolmn' becomes 'MktQty' all over benchTradePerf()

This name is more consistent with similar quantities in blotter, such as 'TxnQty'
